src/telemetry/step_stats.rs

```diff
@@ -0,0 +1,86 @@
+//!
+//! Export metrics collected during protocol run in CSV format. Metrics are partitioned by step.
+
+use crate::telemetry::labels;
+use crate::telemetry::metrics::{INDEXED_PRSS_GENERATED, RECORDS_SENT, SEQUENTIAL_PRSS_GENERATED};
+use crate::telemetry::stats::Metrics;
+use std::collections::{BTreeMap, HashMap};
+use std::io;
+use std::io::{Error, Write};
+
+pub trait CsvExporter {
+    /// Writes the serialized version of this instance into the provided writer in CSV format.
+    ///
+    /// ## Errors
+    /// Returns an error if an IO error occurs while writing to `W`.
+    fn export<W: io::Write>(&self, w: &mut W) -> Result<(), io::Error>;
+}
+
+impl CsvExporter for Metrics {
+    fn export<W: Write>(&self, w: &mut W) -> Result<(), Error> {
+        // first thing is to collect all the steps and metrics emitted
+        let mut steps_stats = StepsStats::default();
+        for (counter_name, details) in &self.counters {
+            if let Some(steps) = details.dimensions.get(labels::STEP) {
+                for (step, val) in steps {
+                    steps_stats.offer(step, counter_name.as_str(), *val);
+                }
+            }
+        }
+
+        // then dump them to the provided Write interface
+        writeln!(w, "Step,Records sent,Indexed PRSS,Sequential PRSS")?;
+        for (step, stats) in steps_stats.all_steps() {
+            writeln!(
+                w,
+                "{},{},{},{}",
+                step,
+                stats.get(RECORDS_SENT),
+                stats.get(INDEXED_PRSS_GENERATED),
+                stats.get(SEQUENTIAL_PRSS_GENERATED)
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+struct StepsStats<'a> {
+    inner: BTreeMap<&'a str, StepStats<'a>>,
+}
+
+impl<'a> StepsStats<'a> {
+    pub fn offer(&mut self, step_name: &'a str, metric: &'a str, val: u64) {
+        if let Some(step_stats) = self.inner.get_mut(step_name) {
```

Comment:
Written in this fashion to convince borrow checker I am not doing anything crazy.